software/firmware/europi.py

```diff
@@ -472,21 +474,57 @@ def __init__(
                 )
         super().__init__(self.width, self.height, i2c)
 
-    def centre_text(self, text):
-        """Split the provided text across 3 lines of display."""
+    def writer(self, font):
+        """Returns the large font writer for the specified font."""
+        n = font.__name__
+        if n not in self.writers:
+            self.writers[n] = Writer(self, font)
+        return self.writers[n]
+
+    def text_len(self, s, font=None):
+        """Returns the length of the string s in pixels."""
+        if font is None:
+            return len(s) * 8
+        else:
+            return self.writer(font).string_len(s)
+
+    def text(self, s, x, y, c=1, font=None):
+        """Display the string s using the x, y coordinates as the upper-left corner of the text.
+        When using a custom font, the text is not displayed if it does not fit in the display.
+        """
+        if font is None:
+            return super().text(s, x, y, c)
+        self.writer(font).print(s, x, y, c)
+
+    def centre_text(self, text, font=None):
+        """Split the provided text across up to 3 lines of display.
+        If a font module reference is passed as parameter, this font is used
+        instead of the default 8x8 font. If the text rendered with the specified
+        font is too large for the display, it will not be displayed.
+        """
         self.fill(0)
         # Default font is 8x8 pixel monospaced font which can be split to a
         # maximum of 4 lines on a 128x32 display, but we limit it to 3 lines
         # for readability.
         lines = str(text).split("\n")
-        maximum_lines = round(self.height / CHAR_HEIGHT)
+        if font is None:
+            h = CHAR_HEIGHT + 1

```

Comment:
[discussion] What does this change mean for font related constants like `CHAR_HEIGHT`? I think that it should probably stay as is, but I could see some arguments against.

Reply:
I don't really understand why the current code is `maximum_lines = round(self.height / CHAR_HEIGHT)` which give 4 as a result although we want only max 3 lines. 

I think the code should be `self.height // (CHAR_HEIGHT + 1)` because later on the value 9 (8+1) is used as the font height.

My code is just a way of solving this. It doesn't change the behavior of the method when using the regular 8x8 font.

Of course, it's trival to update the centre_text() method to keep the current behavior with the 8x8 font and to add the support for custom fonts : 

    def centre_text(self, text, font=None):
        if font is None:
            # current code
        else:  
            # new code for the large fonts

Again, a new DisplayCustom class would solve this easily and garantees that all existing code will continue to execute with no regression at all.

Reply:
If we have a new dedicated class for the custom font, then we can revert to the original code for the default Display class and this CHAR_HEIGHT constant keeps its meaning.